### ISSUE TITLE
Weighted sum/mean/cumsum

### ISSUE THREAD

**Issue by jbusecke** [xgcm/xgcm#102]
While thinking about #99, I would also consider adding a module for weighted mean/sum/cumsum to xgcm. I have been using my own code so far (https://github.com/jbusecke/xarrayutils/blob/master/xarrayutils/weighted_operations.py). 

Currently I would use someting like this 
```
from xarrayutils.weighted_ops import weighted_mean
w_mean = weighted_mean(da['varname'], da['weightname'], dims=['xc', 'yc'])
```
I would really like if I could do something like this in xgcm with a full dataset:
```
from xgcm import Grid
grid = Grid(ds)

ds_mean = grid.mean(['X', 'Y'], weights='area')
```
This would use the internal logic resulting from #99 and weigh all variables with the appropriate area depening on its dimensions.

@rabernat, would you consider this something that could make it into xgcm? This would obviously not be the first thing on the list, but I wanted to put it out here.

**Comment by jbusecke** [comment]
@rabernat layed the foundation to make xgcm 'metric-aware' in #99.
This represents a great step forward to #81.

After discussing things with @rabernat I implemented `Grid.integrate` in #130.
Note that I discovered a bug in #99 which i fixed in this PR and also added a bunch more tests. 

@rabernat, could you take a look at the changes? I might have butchered your elegant function, but for now it does what it is supposed to do. What else do we need to merge this functionality?

- Defintitely nice documentation of the new features.
- Maybe some more tests?

Let me know what you think.

**Comment by jbusecke** [comment]
While the last kinks of `integrate` are ironed out (#130), I would like to start a discussion about other functions using the new metrics capability of xgcm (#136).

The obvious next candidate is a weighted average (I think it should work as `grid.average(da, axis)`.

What other functions could be needed (not that I want to implement them all right now, but its good to discuss):
- weighted std?
- weighted cumsum? (Probably very helpful to calculate streamfunctions?)
- any others?

Are these things that people are looking for? Or too specific for xgcm?

cc @rabernat @dhruvbalwada @roxyboy @anirban89 @raphaeldussin

**Comment by dhruvbalwada** [comment]
Is this similar to what https://xhistogram.readthedocs.io/en/latest/ wants to do in the longterm? Maybe that will be a better project to have these discussions around?

**Comment by rabernat** [comment]
@dhruvbalwada - thanks for weighing in! (No pun intended 🤣 )

Here we are talking about reductions operations happening in grid space. Like a zonal mean or mean over the domain. It's something we already do every day when we do
```
(ds.SST * ds.rA).sum()
```
(area-weighted sum of SST)

The "weight" concept in xhistogram is about something related but different: what weights to assign to each point when calculating a histogram. The weights themselves will indeed often be the same arrays.

We could potentially attach histogram methods to xgcm Grids which would automatically figure out the weights.

**Comment by jbusecke** [comment]
I think this functionality is essential to have for the upcoming CMIP6 hackathon. I am happy to advance this once we have dealt with #154, #130 and #132. Just wanted to ping the issue.

**Comment by jbusecke** [comment]
I have a prototype implementation up in #162.

**Comment by jbusecke** [comment]
Closed via #162
